tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15320,17 +15421,23 @@ void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
       HLSLShaderAttr::CreateImplicit(S.Context, fullName);
 
   FD->addAttr(pShaderAttr);
+  isActiveEntry = true;
   return;
 }
 
 void DiagnoseEntry(Sema &S, FunctionDecl *FD) {
-  TryAddShaderAttrFromTargetProfile(S, FD);
+  bool isActiveEntry = false;
+  TryAddShaderAttrFromTargetProfile(S, FD, isActiveEntry);
 
-  auto Attr = FD->getAttr<HLSLShaderAttr>();
+  HLSLShaderAttr *Attr = FD->getAttr<HLSLShaderAttr>();
   if (!Attr) {
     return;
   }
 
+  if (isActiveEntry) {
+    DiagnoseEntryAttrAllowedOnStage(&S, FD, Attr);
```

Comment:
If this is called after `DXIL::ShaderKind Stage = ShaderModel::KindFromFullName(Attr->getStage());`, you can pass Stage in here, and switch on the enum instead of using StringSwitch there.